### PR TITLE
Add support for Tag metadata

### PR DESCRIPTION
**Pull request by rghetia**
Replaces #1032

### CODE REVIEW

File: tag/map.go [hunk 1]
```diff
@@ -95,13 +100,14 @@ type Mutator interface {
 // Insert returns a mutator that inserts a
 // value associated with k. If k already exists in the tag map,
 // mutator doesn't update the value.
-func Insert(k Key, v string) Mutator {
+// Metadata applies metadata to the tag. It is optional.
```
Comment (songy23) [inline tag/map.go]: What if there're more than one metadatas and they're conflicting? E.g `Insert(k, v, unlimitedPropagation, noPropagation)`?

Comment (rghetia) [reply tag/map.go]: the last one will be effective.

Comment (songy23) [reply tag/map.go]: Consider adding a note on the API document, and adding a few tests on this case.

Comment (rghetia) [reply tag/map.go]: added a note in the apis and testcases.

**Pull request merged**